repose-aggregator/components/services/http-logging-service/http-logging-service-api/src/main/java/org/openrepose/core/services/httplogging/HttpLoggingContext.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * _=_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_=
+ * Repose
+ * _-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-
+ * Copyright (C) 2010 - 2015 Rackspace US, Inc.
+ * _-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * =_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_=_
+ */
+package org.openrepose.core.services.httplogging;
+
+import lombok.AccessLevel;
+import lombok.Data;
+import lombok.RequiredArgsConstructor;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * A container for all of the state associated with an HTTP interaction that
+ * may be used by the {@link HttpLoggingService} to render messages.
+ * <p>
+ * This class is intentionally mutable to keep interactions simple.
+ * Additionally, standard setters are used rather than "fluent" setters to
+ * indicate mutability and keep the behavior of methods clear.
+ * <p>
+ * This class should only be constructed by the {@link HttpLoggingService}.
+ * By only allowing the {@link HttpLoggingService} to construct this class,
+ * stronger assertions can be made about the tracking and integrity of any
+ * instances. For that reason, constructor access is limited to the package.
+ */
+@Data
+@RequiredArgsConstructor(access = AccessLevel.PACKAGE)
+public class HttpLoggingContext {
```

Comment:
This has remained a design problem for me. On the one hand, we need to pass a `Map` to JTwig. A `Map` is a good structure for associating tokens to data. On the other hand, to accommodate the breadth of data we capture, that `Map` would have to be fairly general (e.g., `Map<String, Object>`) which means we would lose type safety which makes working with this context a little more painful.

The current solution was to use a type-safe context object to capture data, then to convert that object to a `Map` to render a template. 

I believe that perhaps a better solution would be to have this context `implements Map<String, Object>`. The context would then delegate to an underlying `HashMap`. What about the type-safety, you ask? The context would add methods to access named, typed data (e.g., `public String getInboundResponseProtocol()`, `public void setInboundResponseProcotol(String protocl)`). That was we don't have to go through a conversion process, we simply build the correct `Map` instead. The context would also perform any necessary mapping (i.e., a method would take a request object, and map it to the corresponding fields like query string, headers, etc.). Extensions could be made directly against the underlying map, or to protect the data we are populating, we could continue to map the key `extensions` to a sub-map and mutate that.

Reply:
I'm on board but would probably suggest composition instead of inheritance in this case.

Reply:
I'm on board but would probably suggest composition instead of inheritance in this case.

Reply:
Funny enough, the more I think about that approach, the more it starts to look like what we already have. I'm not too concerned with making this change right now unless someone has a clear vision of a better solution.